### PR TITLE
Don't do zero length reads/writes on TCP

### PR DESCRIPTION
**Pull request by erikcorry**

### CODE REVIEW

**Comment by erikcorry** [comment]
This goes back to c22e65d9fa

File: lib/net/modules/tcp.toit [hunk 1]
```diff
@@ -152,20 +152,21 @@ class TcpSocket extends TcpSocket_ implements net.Socket Reader:
       throw error
 
   read:
```
Comment (kasperl) [inline lib/net/modules/tcp.toit:154]:
```suggestion
  read -> ByteArray?:
```

File: lib/net/modules/tcp.toit [hunk 2]
```diff
@@ -152,20 +152,21 @@ class TcpSocket extends TcpSocket_ implements net.Socket Reader:
...
-    state.clear_state TOIT_TCP_READ_
-    return ByteArray 0
+    while true:
+      state := ensure_state_ TOIT_TCP_READ_ --failure=: throw it
+      result := tcp_read_ state.group state.resource
+      if result != -1: return result
+      // TODO(anders): We could consider always clearing this after all reads.
+      state.clear_state TOIT_TCP_READ_
 
   write data from = 0 to = data.size:
```
Comment (kasperl) [inline lib/net/modules/tcp.toit:162]:
```suggestion
  write data from = 0 to = data.size -> int:
```

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: lib/net/modules/tcp.toit [hunk 3]
```diff
@@ -151,21 +151,22 @@ class TcpSocket extends TcpSocket_ implements net.Socket Reader:
...
+      state.clear_state TOIT_TCP_READ_
+
+  write data from = 0 to = data.size -> int:
+    while true:
+      state := ensure_state_ TOIT_TCP_WRITE_ --error_bits=(TOIT_TCP_ERROR_ | TOIT_TCP_CLOSE_) --failure=: throw it
+      wrote := tcp_write_ state.group state.resource data from to
+      if wrote == -1:
+        state.clear_state TOIT_TCP_WRITE_
+      else:
+        return wrote
```
Comment (kasperl) [inline lib/net/modules/tcp.toit:169]: Why not write this like in read?

```suggestion
      if wrote != -1: return wrote
      state.clear_state TOIT_TCP_WRITE_
```

**Review by kasperl — APPROVED** [review]
Still looks good with a suggestion:

**Pull request merged**